tests/project_helpers.py

```diff
@@ -0,0 +1,245 @@
+
+    @classmethod
+    def parse_from_toml(cls, data: TomlData) -> "AnalysisExpectations":
+        """Read expectations from the given TOML table."""
+
+        def set_or_none(data: Optional[Iterable[str]]) -> Optional[Set[str]]:
+            return None if data is None else set(data)
+
+        return cls(
+            imports=set_or_none(data.get("imports")),
+            declared_deps=set_or_none(data.get("declared_deps")),
+            undeclared_deps=set_or_none(data.get("undeclared_deps")),
+            unused_deps=set_or_none(data.get("unused_deps")),
+        )
+
+    def _verify_fields(self, member_extractor: Callable[[str], Set[str]]) -> None:
+        for field in fields(self):
+            expected_names = getattr(self, field.name)
+            if expected_names is None:
+                print(f"Skip checking .{field.name}")
+                continue
+
+            print(f"Checking .{field.name}")
+            actual_names = member_extractor(field.name)
+            print(f"  Expect names: {sorted(expected_names)}")
+            print(f"  Actual names: {sorted(actual_names)}")
+            assert expected_names == actual_names
+
+    def verify_analysis(self, analysis: Analysis) -> None:
+        """Assert that the given Analysis object matches our expectations."""
+        self._verify_fields(lambda member: {m.name for m in getattr(analysis, member)})
+
+    def verify_analysis_json(self, analysis: JsonData) -> None:
+        """Assert that the given JSON analysis matches our expectations."""
+        self._verify_fields(lambda member: {m["name"] for m in analysis[member]})
+
+
+@dataclass
+class BaseExperiment:
+    """A single experiment, running FawltyDeps on a test project.
+
+    An experiment is part of a bigger project (see BaseProject below) and has:
+    - A name and description, for documentation purposes.
+    - A list of requirements, to be installed into a virtualenv and made
+      available to FawltyDeps when this experiment is run
+      (see CachedExperimentVenv for details).
+    - A set of expectations on the resulting Analysis object, to be verified
+      after the FawltyDeps has been run (see AnalysisExpectations for details).
+    """
+
+    name: str
+    description: Optional[str]
+    requirements: List[str]
+    expectations: AnalysisExpectations
+
+    @staticmethod
+    def init_args_from_toml(name: str, data: TomlData) -> Dict[str, Any]:
+        return dict(
+            name=name,
+            description=data.get("description"),
+            requirements=data.get("requirements", []),
+            expectations=AnalysisExpectations.parse_from_toml(data),
+        )
+
+    @classmethod
+    def parse_from_toml(cls, name: str, data: TomlData) -> "BaseExperiment":
+        raise NotImplementedError
+
+    def get_venv_dir(self, cache: pytest.Cache) -> Path:
+        """Get this venv's dir and create it if necessary."""
+        venv = CachedExperimentVenv(self.requirements)
+        return venv(cache)
```

Comment:
Nit:
```suggestion
        return CachedExperimentVenv(self.requirements)(cache)
```
if that's allowed syntax-wise

Reply:
It's allowed, but I've found that readers not familiar with FP can often get tripped up by the double parentheses. Then again, if FP is absolutely foreign to them, then I suspect they will have a bad time reading our codebase in any case... 😅

Reply:
fixed in b8865ad